a- Read	the	iris	dataset	from	the	following	URL:	
'https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv'			
and	assign	it	to	a	Pandas	DataFrame	as	you	learned	in	tutorial	Lab2-3.	

In [80]:
from sklearn.datasets import load_iris 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.cross_validation import train_test_split
from sklearn.metrics import accuracy_score 
import pandas as pd
df = pd.DataFrame()
iris_df = pd.read_csv('https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv')


b- Split	 the	 dataset	 into	 testing	 and	 training	 sets	 with	 the	 following	 parameters:	
test_size=0.4,	random_state=1	

In [81]:
iris = load_iris()
X= iris.data
y=iris.target
X_train, X_test,y_train,y_test = train_test_split(X, y, test_size=0.4, random_state=1)



c- Instantiate	a	KNN	object	with	K=3,	train	it	on	the	training	set	and	test	it	on	the	testing	set.	
Then,	calculate	the	accuracy	of	your	prediction	as	you	learned	in	Lab3.

In [82]:
K=3
knn = KNeighborsClassifier(n_neighbors=K)
knn.fit(X_train, y_train)
y_predict = knn.predict(X_test)
accuracy = accuracy_score(y_test, y_predict)

print(score)

0.8


d- Repeat	part	(c)		for	K=1,	K=5,	K=7,	K=15,	K=28,	K=60	(you	can	simply	write	a	“for	loop”	on	
a	list	with	these	values	to	repeat	the	process	for	you,	and	save	the	final	accuracy	results	
in	a	list	(array)!).	Does	the	accuracy	always	get	better	by	increasing	K?		

In [84]:
X_train, X_test,y_train,y_test = train_test_split(X, y, test_size=0.4, random_state=1)    
kList = [1, 5, 7, 15, 28, 60]
accuracyList=[]
for k in kList:
    knn= KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    y_predict = knn.predict(X_test)
    accuracy = accuracy_score(y_test, y_predict)
    accuracyList.append(accuracy)
    
print(accuracyList)

[0.96666666666666667, 0.98333333333333328, 0.98333333333333328, 0.98333333333333328, 0.94999999999999996, 0.80000000000000004]


Does the accuracy always get better by increasing K? 

### It does not look like it's getting better when K=60 its 0.80000000000000004 and  K = 5 is 0.98333333333333328

e- Now,	suppose	that	we	would	like	to	make	prediction	based	on	only	ONE	single	feature.	
To	find	the	best	single	feature,	we	have	to	try	every	feature	individually.	In	other	word,	
we	 have	 to	 repeat	 part	 (c)	 4	 times	 (each	 time	 using	 only	 one	 of	 the	 4	 features),	 and	
compute	the	accuracy	each	time.	Then,	check	the	accuracies.	Which	individual	feature	
provide	 the	 best	 accuracy	 (the	 best	 feature)?	 	 Which	 one	 is	 the	 second	 best	 feature?	
(Note:	you	have	to	train,	test,	and	evaluate	your	model	4	times,	each	time	on	a	dataset	
including	only	one	of	the	features.	You	may	want	to	write	a	“for	loop”	to	generate	the	
datasets,	train,	test,	and	compute	the	accuracy	each	time,	and	save	the	final	accuracy	
results	in	a	list).	

In [85]:
def categorical_to_numeric(x):
    if x == 'setosa':
        return 0
    elif x == 'versicolor':
        return 1
    elif x == 'virginica':
        return 2
    

iris_df['label'] = iris_df['species'].apply(categorical_to_numeric)

iris_df[0::10]



sepal_length  sepal_width  petal_length  petal_width     species  label
0             5.1          3.5           1.4          0.2      setosa      0
10            5.4          3.7           1.5          0.2      setosa      0
20            5.4          3.4           1.7          0.2      setosa      0
30            4.8          3.1           1.6          0.2      setosa      0
40            5.0          3.5           1.3          0.3      setosa      0
50            7.0          3.2           4.7          1.4  versicolor      1
60            5.0          2.0           3.5          1.0  versicolor      1
70            5.9          3.2           4.8          1.8  versicolor      1
80            5.5          2.4           3.8          1.1  versicolor      1
90            5.5          2.6           4.4          1.2  versicolor      1
100           6.3          3.3           6.0          2.5   virginica      2
110           6.5          3.2           5.1          2.0   virginica      2
120           6.9          3.2           5.7          2.3   virginica      2
130           7.4          2.8           6.1          1.9   virginica      2
140           6.7          3.1           5.6          2.4   virginica      2

f- Now,	we	want	to	repeat	part	(e),	this	time	using	two	features.	you	have	to	train,	test,	and	
evaluate	 your	 model	 for	 6	 different	 cases:	 using	 (1 st 	 and	 2 nd 	 features),	 (1 st 	 and	 3 rd 		
features),	(1 st 	and	4 th 		features),	(2 nd 		and	3 rd 		features),	(2 nd 	and	4 th 	features),	(3 rd 	and	4 th 		
features)!																																												
Which “feature	pair”	provide	the	best	accuracy?	

g- Big	Question:	Does	the	“best	feature	pair”	from	part	(f)	contain	of	both	“first	best	feature”	
and	“second	best	feature”	from	part	(e)?		In	other	word,	can	we	conclude	that	the	“best	
two	features”	for	classification	are	the	“first	best	individual	feature”	and	the	“second	best	
individual	feature”	together?	

h- Optional	Question:	Justify	your	answer	to	part	(g)!		If	yes,	why?		If	no,	why	not?	